templates/shift-scheduling/main.go

```diff
@@ -1,370 +1,248 @@
-	// negative as we want to maximize it while minimizing the worker count.
-	workerCountImportance := 10
-	schedule = schedule.Value(func(s store.Store) int {
-		return workerCountImportance*workerCount.Get(s) - happiness.Get(s)
-	})
-
-	// Now we define our constraint containing a propagator. This will remove
-	// workers from domains based upon the workers in other domains.
-	constraint := newBreakConstraint(shifts, 2)
-
-	// We invoke the propagator to propagate the initial values.
-	schedule.Propagate(constraint.propagate)
-
-	if constraint.isProvenInfeasible(schedule) {
-		return nil, fmt.Errorf("input can not be solved")
+	m := mip.NewModel()
+	m.Objective().SetMinimize()
+
+	x := model.NewMultiMap(
+		func(...Assignment) mip.Bool {
+			return m.NewBool()
+		}, potentialAssignments)
+
+	underSupplySlack := model.NewMultiMap(
+		func(demand ...RequiredWorker) mip.Float {
+			return m.NewFloat(0, float64(demand[0].Count))
+		}, input.RequiredWorkers)
+
+	overSupplySlack := model.NewMultiMap(
+		func(demand ...RequiredWorker) mip.Float {
+			return m.NewFloat(0, math.MaxFloat64)
+		}, input.RequiredWorkers)
+
+	for _, demand := range input.RequiredWorkers {
+		demandCover := demandCovering[demand.RequiredWorkerID]
+		// We need to cover all demands
+		coverConstraint := m.NewConstraint(mip.Equal, float64(demand.Count))
+		coverConstraint.NewTerm(1.0, underSupplySlack.Get(demand))
+		coverConstraint.NewTerm(-1.0, overSupplySlack.Get(demand))
+		coverPerDriver := map[int]mip.Constraint{}
+		for _, assignment := range demandCover {
+			if constraint, ok := coverPerDriver[assignment.Worker.ID]; !ok {
+				// When two potential assignments of a driver overlap, only one
+				// can be assigned
+				coverPerDriver[assignment.Worker.ID] = m.NewConstraint(mip.LessThanOrEqual, 1.0)
+				coverPerDriver[assignment.Worker.ID].NewTerm(1.0, x.Get(assignment))
+			} else {
+				constraint.NewTerm(1.0, x.Get(assignment))
+			}
+			coverConstraint.NewTerm(1.0, x.Get(assignment))
+		}
+		m.Objective().NewTerm(opts.OverSupplyPenalty, overSupplySlack.Get(demand))
+		m.Objective().NewTerm(opts.UnderSupplyPenalty, underSupplySlack.Get(demand))
 	}
 
-	// Now the most important part: given a store, how can we create new
-	// stores that bring us closer to a valid schedule.
-	schedule = schedule.Generate(func(s store.Store) store.Generator {
-		// We take the first shift for which a worker has not been yet assigned.
-		// Then we generate for each worker that can be assigned to a shift
-		// i a new store.
-		i, ok := shifts.First(s)
-		workers := shifts.Domain(s, i).Slice()
-
-		return store.Lazy(
-			func() bool {
-				for j := 0; j < nShifts; j++ {
-					if shifts.Domain(s, j).Empty() {
-						return false
+	// Two shift of a driver have to be at least 8 hours apart
+	for _, driver := range input.Workers {
+		for i, a1 := range potentialAssignmentsPerDriver[driver.ID] {
+			// A driver can only work 10 hours per day
```

Comment:
We don't have the concept of drivers in the schema. We should call them workers

Reply:
replaced all notions of driver with worker